issue357 - sitecaptain name is displayed

kate-mills opened the pull request:
You can check out this version at, https://issue357-dot-rebuildingtogethercaptain-hrd.appspot.com/room/staff_home

kate-mills committed:
sitecaptain name is displayed

babybunny commented on gae/js/app/views/sitecaptain.js:
```diff
@@ -68,11 +68,18 @@ define(
+            before_save: function(e){
+                e.preventDefault();
+                var id_value = Number(this.$el.find('select').val());
+                var choices = this.form.fields.models[1].changed.options;
```
The [1] looks brittle - meaning that if something changes about that array this will break suddenly.  Let's find another way to refer to that list of options.

babybunny replied:
Also... I 'm getting a console error sometimes which may be related.  Seems that this.filtered_captain is not defined. Sometimes. But sometimes it works.  

jquery.js:2 jQuery.Deferred exception: Cannot read property 'label' of undefined TypeError: Cannot read property 'label' of undefined
    at Object.<anonymous> (https://issue357-dot-rebuildingtogethercaptain-hrd.appspot.com/js/app/views/sitecaptain.js:82:65)
    at j (https://issue357-dot-rebuildingtogethercaptain-hrd.appspot.com/js/lib/jquery.js:2:29999)
    at k (https://issue357-dot-rebuildingtogethercaptain-hrd.appspot.com/js/lib/jquery.js:2:30313) undefined

babybunny reviewed (commented):
Hi Katie, 

Sorry it took a few days to get to this. Looks like a great approach.  Finding the correct value in the list of choices and extracting the name. 

I'm seeing a JS error in console, and also a server error on /cru_api.sitecaptain_delete.  Are you available to debug it with me?  We could walk through it together.

kate-mills commented:
I would love to debug it with you. I am available tomorrow to debug. What about you?

Sent from my iPhone

> On Nov 11, 2017, at 10:45 AM, Luke Stone <notifications@github.com> wrote:
> 
> @babybunny commented on this pull request.
> 
> Hi Katie,
> 
> Sorry it took a few days to get to this. Looks like a great approach. Finding the correct value in the list of choices and extracting the name.
> 
> I'm seeing a JS error in console, and also a server error on /cru_api.sitecaptain_delete. Are you available to debug it with me? We could walk through it together.
> 
> In gae/js/app/views/sitecaptain.js:
> 
> > @@ -68,11 +68,18 @@ define(
>                  m.destroy();
>                  this.render();
>              },
> +            before_save: function(e){
> +                e.preventDefault();
> +                var id_value = Number(this.$el.find('select').val());
> +                var choices = this.form.fields.models[1].changed.options;
> The [1] looks brittle - meaning that if something changes about that array this will break suddenly. Let's find another way to refer to that list of options.
> 
> In gae/js/app/views/sitecaptain.js:
> 
> > @@ -68,11 +68,18 @@ define(
>                  m.destroy();
>                  this.render();
>              },
> +            before_save: function(e){
> +                e.preventDefault();
> +                var id_value = Number(this.$el.find('select').val());
> +                var choices = this.form.fields.models[1].changed.options;
> Also... I 'm getting a console error sometimes which may be related. Seems that this.filtered_captain is not defined. Sometimes. But sometimes it works.
> 
> jquery.js:2 jQuery.Deferred exception: Cannot read property 'label' of undefined TypeError: Cannot read property 'label' of undefined
> at Object. (https://issue357-dot-rebuildingtogethercaptain-hrd.appspot.com/js/app/views/sitecaptain.js:82:65)
> at j (https://issue357-dot-rebuildingtogethercaptain-hrd.appspot.com/js/lib/jquery.js:2:29999)
> at k (https://issue357-dot-rebuildingtogethercaptain-hrd.appspot.com/js/lib/jquery.js:2:30313) undefined
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

babybunny commented:
Yeah I'm available tomorrow as well.  One to five are good hours.

On Nov 11, 2017 12:22 PM, "Kate Mills" <notifications@github.com> wrote:

> I would love to debug it with you. I am available tomorrow to debug. What
> about you?
>
> Sent from my iPhone
>
> > On Nov 11, 2017, at 10:45 AM, Luke Stone <notifications@github.com>
> wrote:
> >
> > @babybunny commented on this pull request.
> >
> > Hi Katie,
> >
> > Sorry it took a few days to get to this. Looks like a great approach.
> Finding the correct value in the list of choices and extracting the name.
> >
> > I'm seeing a JS error in console, and also a server error on
> /cru_api.sitecaptain_delete. Are you available to debug it with me? We
> could walk through it together.
> >
> > In gae/js/app/views/sitecaptain.js:
> >
> > > @@ -68,11 +68,18 @@ define(
> > m.destroy();
> > this.render();
> > },
> > + before_save: function(e){
> > + e.preventDefault();
> > + var id_value = Number(this.$el.find('select').val());
> > + var choices = this.form.fields.models[1].changed.options;
> > The [1] looks brittle - meaning that if something changes about that
> array this will break suddenly. Let's find another way to refer to that
> list of options.
> >
> > In gae/js/app/views/sitecaptain.js:
> >
> > > @@ -68,11 +68,18 @@ define(
> > m.destroy();
> > this.render();
> > },
> > + before_save: function(e){
> > + e.preventDefault();
> > + var id_value = Number(this.$el.find('select').val());
> > + var choices = this.form.fields.models[1].changed.options;
> > Also... I 'm getting a console error sometimes which may be related.
> Seems that this.filtered_captain is not defined. Sometimes. But sometimes
> it works.
> >
> > jquery.js:2 jQuery.Deferred exception: Cannot read property 'label' of
> undefined TypeError: Cannot read property 'label' of undefined
> > at Object. (https://issue357-dot-rebuildingtogethercaptain-hrd.
> appspot.com/js/app/views/sitecaptain.js:82:65)
> > at j (https://issue357-dot-rebuildingtogethercaptain-hrd.
> appspot.com/js/lib/jquery.js:2:29999)
> > at k (https://issue357-dot-rebuildingtogethercaptain-hrd.
> appspot.com/js/lib/jquery.js:2:30313) undefined
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/babybunny/rebuildingtogethercaptain/pull/405#issuecomment-343691550>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AK-yPHwDTcwK8nafqd3TNnCd-xRekEvqks5s1gIPgaJpZM4QXXCE>
> .
>

kate-mills commented:
Perfect!  Im available at 1:00 as well. 

Sent from my iPhone

> On Nov 11, 2017, at 12:57 PM, Luke Stone <notifications@github.com> wrote:
> 
> Yeah I'm available tomorrow as well. One to five are good hours.
> 
> On Nov 11, 2017 12:22 PM, "Kate Mills" <notifications@github.com> wrote:
> 
> > I would love to debug it with you. I am available tomorrow to debug. What
> > about you?
> >
> > Sent from my iPhone
> >
> > > On Nov 11, 2017, at 10:45 AM, Luke Stone <notifications@github.com>
> > wrote:
> > >
> > > @babybunny commented on this pull request.
> > >
> > > Hi Katie,
> > >
> > > Sorry it took a few days to get to this. Looks like a great approach.
> > Finding the correct value in the list of choices and extracting the name.
> > >
> > > I'm seeing a JS error in console, and also a server error on
> > /cru_api.sitecaptain_delete. Are you available to debug it with me? We
> > could walk through it together.
> > >
> > > In gae/js/app/views/sitecaptain.js:
> > >
> > > > @@ -68,11 +68,18 @@ define(
> > > m.destroy();
> > > this.render();
> > > },
> > > + before_save: function(e){
> > > + e.preventDefault();
> > > + var id_value = Number(this.$el.find('select').val());
> > > + var choices = this.form.fields.models[1].changed.options;
> > > The [1] looks brittle - meaning that if something changes about that
> > array this will break suddenly. Let's find another way to refer to that
> > list of options.
> > >
> > > In gae/js/app/views/sitecaptain.js:
> > >
> > > > @@ -68,11 +68,18 @@ define(
> > > m.destroy();
> > > this.render();
> > > },
> > > + before_save: function(e){
> > > + e.preventDefault();
> > > + var id_value = Number(this.$el.find('select').val());
> > > + var choices = this.form.fields.models[1].changed.options;
> > > Also... I 'm getting a console error sometimes which may be related.
> > Seems that this.filtered_captain is not defined. Sometimes. But sometimes
> > it works.
> > >
> > > jquery.js:2 jQuery.Deferred exception: Cannot read property 'label' of
> > undefined TypeError: Cannot read property 'label' of undefined
> > > at Object. (https://issue357-dot-rebuildingtogethercaptain-hrd.
> > appspot.com/js/app/views/sitecaptain.js:82:65)
> > > at j (https://issue357-dot-rebuildingtogethercaptain-hrd.
> > appspot.com/js/lib/jquery.js:2:29999)
> > > at k (https://issue357-dot-rebuildingtogethercaptain-hrd.
> > appspot.com/js/lib/jquery.js:2:30313) undefined
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub, or mute the thread.
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/babybunny/rebuildingtogethercaptain/pull/405#issuecomment-343691550>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AK-yPHwDTcwK8nafqd3TNnCd-xRekEvqks5s1gIPgaJpZM4QXXCE>
> > .
> >
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

babybunny commented:
More like one thirty I guess

On Nov 11, 2017 1:24 PM, "Kate Mills" <notifications@github.com> wrote:

> Perfect! Im available at 1:00 as well.
>
> Sent from my iPhone
>
> > On Nov 11, 2017, at 12:57 PM, Luke Stone <notifications@github.com>
> wrote:
> >
> > Yeah I'm available tomorrow as well. One to five are good hours.
> >
> > On Nov 11, 2017 12:22 PM, "Kate Mills" <notifications@github.com> wrote:
> >
> > > I would love to debug it with you. I am available tomorrow to debug.
> What
> > > about you?
> > >
> > > Sent from my iPhone
> > >
> > > > On Nov 11, 2017, at 10:45 AM, Luke Stone <notifications@github.com>
> > > wrote:
> > > >
> > > > @babybunny commented on this pull request.
> > > >
> > > > Hi Katie,
> > > >
> > > > Sorry it took a few days to get to this. Looks like a great approach.
> > > Finding the correct value in the list of choices and extracting the
> name.
> > > >
> > > > I'm seeing a JS error in console, and also a server error on
> > > /cru_api.sitecaptain_delete. Are you available to debug it with me? We
> > > could walk through it together.
> > > >
> > > > In gae/js/app/views/sitecaptain.js:
> > > >
> > > > > @@ -68,11 +68,18 @@ define(
> > > > m.destroy();
> > > > this.render();
> > > > },
> > > > + before_save: function(e){
> > > > + e.preventDefault();
> > > > + var id_value = Number(this.$el.find('select').val());
> > > > + var choices = this.form.fields.models[1].changed.options;
> > > > The [1] looks brittle - meaning that if something changes about that
> > > array this will break suddenly. Let's find another way to refer to that
> > > list of options.
> > > >
> > > > In gae/js/app/views/sitecaptain.js:
> > > >
> > > > > @@ -68,11 +68,18 @@ define(
> > > > m.destroy();
> > > > this.render();
> > > > },
> > > > + before_save: function(e){
> > > > + e.preventDefault();
> > > > + var id_value = Number(this.$el.find('select').val());
> > > > + var choices = this.form.fields.models[1].changed.options;
> > > > Also... I 'm getting a console error sometimes which may be related.
> > > Seems that this.filtered_captain is not defined. Sometimes. But
> sometimes
> > > it works.
> > > >
> > > > jquery.js:2 jQuery.Deferred exception: Cannot read property 'label'
> of
> > > undefined TypeError: Cannot read property 'label' of undefined
> > > > at Object. (https://issue357-dot-rebuildingtogethercaptain-hrd.
> > > appspot.com/js/app/views/sitecaptain.js:82:65)
> > > > at j (https://issue357-dot-rebuildingtogethercaptain-hrd.
> > > appspot.com/js/lib/jquery.js:2:29999)
> > > > at k (https://issue357-dot-rebuildingtogethercaptain-hrd.
> > > appspot.com/js/lib/jquery.js:2:30313) undefined
> > > >
> > > > —
> > > > You are receiving this because you authored the thread.
> > > > Reply to this email directly, view it on GitHub, or mute the thread.
> > > >
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/babybunny/rebuildingtogethercaptain/
> pull/405#issuecomment-343691550>,
> > > or mute the thread
> > > <https://github.com/notifications/unsubscribe-
> auth/AK-yPHwDTcwK8nafqd3TNnCd-xRekEvqks5s1gIPgaJpZM4QXXCE>
> > > .
> > >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/babybunny/rebuildingtogethercaptain/pull/405#issuecomment-343695161>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AK-yPM0Fmu9sS03mp12-_E1eklpI-L_yks5s1hCSgaJpZM4QXXCE>
> .
>

kate-mills commented:
np. ping me when you are avail.

On Sun, Nov 12, 2017 at 12:42 PM, Luke Stone <notifications@github.com>
wrote:

> More like one thirty I guess
>
>
> On Nov 11, 2017 1:24 PM, "Kate Mills" <notifications@github.com> wrote:
>
> > Perfect! Im available at 1:00 as well.
> >
> > Sent from my iPhone
> >
> > > On Nov 11, 2017, at 12:57 PM, Luke Stone <notifications@github.com>
> > wrote:
> > >
> > > Yeah I'm available tomorrow as well. One to five are good hours.
> > >
> > > On Nov 11, 2017 12:22 PM, "Kate Mills" <notifications@github.com>
> wrote:
> > >
> > > > I would love to debug it with you. I am available tomorrow to debug.
> > What
> > > > about you?
> > > >
> > > > Sent from my iPhone
> > > >
> > > > > On Nov 11, 2017, at 10:45 AM, Luke Stone <notifications@github.com
> >
> > > > wrote:
> > > > >
> > > > > @babybunny commented on this pull request.
> > > > >
> > > > > Hi Katie,
> > > > >
> > > > > Sorry it took a few days to get to this. Looks like a great
> approach.
> > > > Finding the correct value in the list of choices and extracting the
> > name.
> > > > >
> > > > > I'm seeing a JS error in console, and also a server error on
> > > > /cru_api.sitecaptain_delete. Are you available to debug it with me?
> We
> > > > could walk through it together.
> > > > >
> > > > > In gae/js/app/views/sitecaptain.js:
> > > > >
> > > > > > @@ -68,11 +68,18 @@ define(
> > > > > m.destroy();
> > > > > this.render();
> > > > > },
> > > > > + before_save: function(e){
> > > > > + e.preventDefault();
> > > > > + var id_value = Number(this.$el.find('select').val());
> > > > > + var choices = this.form.fields.models[1].changed.options;
> > > > > The [1] looks brittle - meaning that if something changes about
> that
> > > > array this will break suddenly. Let's find another way to refer to
> that
> > > > list of options.
> > > > >
> > > > > In gae/js/app/views/sitecaptain.js:
> > > > >
> > > > > > @@ -68,11 +68,18 @@ define(
> > > > > m.destroy();
> > > > > this.render();
> > > > > },
> > > > > + before_save: function(e){
> > > > > + e.preventDefault();
> > > > > + var id_value = Number(this.$el.find('select').val());
> > > > > + var choices = this.form.fields.models[1].changed.options;
> > > > > Also... I 'm getting a console error sometimes which may be
> related.
> > > > Seems that this.filtered_captain is not defined. Sometimes. But
> > sometimes
> > > > it works.
> > > > >
> > > > > jquery.js:2 jQuery.Deferred exception: Cannot read property 'label'
> > of
> > > > undefined TypeError: Cannot read property 'label' of undefined
> > > > > at Object. (https://issue357-dot-rebuildingtogethercaptain-hrd.
> > > > appspot.com/js/app/views/sitecaptain.js:82:65)
> > > > > at j (https://issue357-dot-rebuildingtogethercaptain-hrd.
> > > > appspot.com/js/lib/jquery.js:2:29999)
> > > > > at k (https://issue357-dot-rebuildingtogethercaptain-hrd.
> > > > appspot.com/js/lib/jquery.js:2:30313) undefined
> > > > >
> > > > > —
> > > > > You are receiving this because you authored the thread.
> > > > > Reply to this email directly, view it on GitHub, or mute the
> thread.
> > > > >
> > > >
> > > > —
> > > > You are receiving this because you were mentioned.
> > > > Reply to this email directly, view it on GitHub
> > > > <https://github.com/babybunny/rebuildingtogethercaptain/
> > pull/405#issuecomment-343691550>,
> > > > or mute the thread
> > > > <https://github.com/notifications/unsubscribe-
> > auth/AK-yPHwDTcwK8nafqd3TNnCd-xRekEvqks5s1gIPgaJpZM4QXXCE>
> > > > .
> > > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub, or mute the thread.
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/babybunny/rebuildingtogethercaptain/
> pull/405#issuecomment-343695161>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AK-yPM0Fmu9sS03mp12-_
> E1eklpI-L_yks5s1hCSgaJpZM4QXXCE>
>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/babybunny/rebuildingtogethercaptain/pull/405#issuecomment-343766008>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APciz1vwbvvpHBldsfQZpHu5VM7CNzFJks5s11gfgaJpZM4QXXCE>
> .
>

kate-mills committed:
removed before_save into addCaptain

babybunny commented on gae/js/app/views/sitecaptain.js:
```diff
@@ -68,11 +68,15 @@ define(
+            addCaptain: function(e) {
+                e.preventDefault();
                 var self = this;
+                var choices = this.form.fields.models.find(function(model){return model.attributes.name == 'captain';});
```
.get('name')

babybunny commented on gae/js/app/views/sitecaptain.js:
```diff
@@ -68,11 +68,15 @@ define(
+                e.preventDefault();
                 var self = this;
+                var choices = this.form.fields.models.find(function(model){return model.attributes.name == 'captain';});
+                var captain_id = this.model.attributes.captain;
```
.get() here too

babybunny commented on gae/js/app/views/sitecaptain.js:
```diff
@@ -68,11 +68,15 @@ define(
                 var self = this;
+                var choices = this.form.fields.models.find(function(model){return model.attributes.name == 'captain';});
+                var captain_id = this.model.attributes.captain;
+                this.captain  = _.findWhere(choices.changed.options, {value: captain_id});
```
this is choices.attributes.get('options')

this.captain can be a variable used in the closure below ("var captain = " .... )

kate-mills committed:
Using model.get vs model.attributes

kate-mills committed:
Merge branch 'service' of https://github.com/babybunny/rebuildingtogethercaptain into issue357

kate-mills commented on gae/js/app/views/sitecaptain.js:
```diff
@@ -71,12 +71,12 @@ define(
+                var choices = this.form.fields.models.find(function(model){return model.get('name') == 'captain';});
+                var captain_id = this.model.get('captain');
+                var captain  = _.findWhere(choices.get('options'), {value: captain_id});
 
```
Do I really need to declare var captain_id?
I declared it for readability but don't like it anymore.
What do you think about passing {value: this.model.get('captain')} instead?

kate-mills commented on gae/js/app/views/sitecaptain.js:
```diff
@@ -71,12 +71,12 @@ define(
-                this.captain  = _.findWhere(choices.changed.options, {value: captain_id});
+                var choices = this.form.fields.models.find(function(model){return model.get('name') == 'captain';});
+                var captain_id = this.model.get('captain');
+                var captain  = _.findWhere(choices.get('options'), {value: captain_id});
```
Do I really need to declare var captain_id?
I declared it for readability but don't like it anymore.
What do you think about passing {value: this.model.get('captain')} into _.findWhere  instead?

babybunny replied:
You're right, var captain_id is not needed and passing this.model.get('captain') into findwhere is perfectly readable to me.

babybunny commented on gae/js/app/views/sitecaptain.js:
```diff
@@ -68,11 +68,15 @@ define(
                 var self = this;
+                var choices = this.form.fields.models.find(function(model){return model.get('name') == 'captain';});
+                var captain_id = this.model.get('captain');
+                var captain  = _.findWhere(choices.get('options'), {value: captain_id});
```
this is where you need to grab the captain label.  then it's a string and you have a safe copy.  the reference to captain is an object and could change before the callback is invoked. 

var captain_label = ..... .label;

kate-mills replied:
I GET it !!!! You are so patient Luke!!

kate-mills committed:
captain_label is a string

kate-mills commented on gae/js/app/views/sitecaptain.js:
```diff
@@ -68,11 +68,15 @@ define(
                 var self = this;
+                var choices = this.form.fields.models.find(function(model){return model.get('name') == 'captain';});
+                var captain_id = this.model.get('captain');
+                var captain  = _.findWhere(choices.get('options'), {value: captain_id});
```
You can see the changes here: https://issue357-dot-rebuildingtogethercaptain-hrd.appspot.com/room/staff_home

kate-mills committed:
replaced save.then with save.success

babybunny reviewed (approved):
Looks good. 

I think it would also work as 
var captain_label = _.findWhere(choices.get('options'), {value: self.model.get('captain')}).label;

kate-mills committed:
captain_label found in 1 line

kate-mills committed:
Merge branch 'service' of https://github.com/babybunny/rebuildingtogethercaptain into issue357

kate-mills commented:
Getting captain_label in 1 line is better than 3, so I made the change.
Also, I fixed the merge conflicts and will merge into service now!


On Fri, Dec 1, 2017 at 12:37 AM, Luke Stone <notifications@github.com>
wrote:

> *@babybunny* approved this pull request.
>
> Looks good.
>
> I think it would also work as
> var captain_label = _.findWhere(choices.get('options'), {value:
> self.model.get('captain')}).label;
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/babybunny/rebuildingtogethercaptain/pull/405#pullrequestreview-80431471>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APciz5NDs-33PftS3ZSYl3GV1d_93rkSks5s77rYgaJpZM4QXXCE>
> .
>
